### PR TITLE
Add `PopupCloseBehavior`

### PR DESCRIPTION
**Pull request by Umatriz**
This PR adds `PopupCloseBehavior` to improve state of the <https://github.com/emilk/egui/issues/4607>

`PopupCloseBehavior` determines when popup will be closed.
- `CloseOnClick` popup will be closed if the click happens anywhere even in the popup's body
- `CloseOnClickAway` popup will be closed if the click happens somewhere else but in the popup's body.

It also adds a test in the demo app which contains several popups examples.

---

My ideas about <https://github.com/emilk/egui/issues/4607> is to make every tooltip and popup a menu. So it will provide more control over popups and tooltips (you will be able to close a popup by calling something similar to the `ui.close_menu` if you need to). You won't need to manually handle it's opening. And also will allow to have multiple popups opened. That means you can have a popup inside a popup. And it will also lead to the easier creation of the popups. (should we create a tracking issue to track changes because to me it seems like a huge amount of changes to be done?)

---

- Improvements on <https://github.com/emilk/egui/issues/4607>

### CODE REVIEW

File: crates/egui/src/containers/popup.rs [hunk 1]
```diff
@@ -253,18 +253,32 @@ pub fn was_tooltip_open_last_frame(ctx: &Context, widget_id: Id) -> bool {
     })
 }
 
+/// Determines popup's close behavior
+#[derive(Clone, Copy)]
+pub enum PopupCloseBehavior {
+    /// Popup will be closed on click
+    /// It is used in the [`ComboBox`]
```
Comment (emilk) [inline crates/egui/src/containers/popup.rs:260]:
```suggestion
    /// Popup will be closed on click anywhere, inside or outside the popup.
    ///
    /// It is used in [`ComboBox`].
```

File: crates/egui/src/containers/popup.rs [hunk 2]
```diff
@@ -253,18 +253,32 @@ pub fn was_tooltip_open_last_frame(ctx: &Context, widget_id: Id) -> bool {
     })
 }
 
+/// Determines popup's close behavior
+#[derive(Clone, Copy)]
+pub enum PopupCloseBehavior {
```
Comment (emilk) [inline crates/egui/src/containers/popup.rs:258]: For completeness, I suggest we also add a third alternative called e.g. `IgnoreClicks` which ignores any clicks. You can only close it by pressing Escape or calling `Memory::close_popup` (and maybe later with `Ui::close_menu`).

----

Alternatively we replace this `enum PopupCloseBehavior` with 

```rs
struct PopupOptions {
    pub close_on_click_inside: bool,
    pub close_on_click_outside: bool,
    pub close_on_escape: bool,
}

impl PopupOptions {
    pub const CLOSE_ON_CLICK_ANYWHERE: Self = …;
    pub const CLOSE_ON_CLICK_INSIDE: Self = …;
}
```

That makes it much easier to expand with more options in the future

File: crates/egui/src/containers/popup.rs [hunk 3]
```diff
@@ -253,18 +253,32 @@ pub fn was_tooltip_open_last_frame(ctx: &Context, widget_id: Id) -> bool {
...
+    /// Popup will be closed on click
+    /// It is used in the [`ComboBox`]
+    CloseOnClick,
+
+    /// Popup will be closed if the click happened somewhere else
+    /// but in the popup's body
+    CloseOnClickAway,
```
Comment (emilk) [inline crates/egui/src/containers/popup.rs:265]:
```suggestion
    CloseOnClickOutside,
```

File: crates/egui/src/containers/popup.rs [hunk 4]
```diff
@@ -327,13 +342,19 @@ pub fn popup_above_or_below_widget<R>(
                         .inner
                     })
                     .inner
-            })
-            .inner;
+            });
+
+        let is_close = match close_behavior {
```
Comment (emilk) [inline crates/egui/src/containers/popup.rs:347]:
```suggestion
        let should_close = match close_behavior {
```

("is close" sounds like it is about proximity)

**Review by emilk — CHANGES_REQUESTED** [review]
Good idea!

**Review by emilk — APPROVED** [review]

**Comment by emilk** [comment]
CI is red though

**Comment by Umatriz** [comment]
@emilk I fixed the doc test so now everything should be fine

**Pull request merged**